### ISSUE TITLE
Out of memory: Kill process ### (rsyslogd) score 727 or sacrifice child

### ISSUE THREAD

**Issue by mike1li** [rsyslog/rsyslog#4017]
### Expected behavior
No "out of memory"
### Actual behavior
from "dmesg -e"
[  +0.007186] Out of memory: Kill process 9057 (rsyslogd) score 716 or sacrifice child
[  +0.006766] Killed process 9057 (rsyslogd) total-vm:3475172kB, anon-rss:1474340kB, file-rss:0kB, shmem-rss:0kB
[  +0.137045] oom_reaper: reaped process 9057 (rsyslogd), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
[Dec 3 03:34] Process accounting resumed

[  +0.007188] Out of memory: Kill process 19010 (rsyslogd) score 726 or sacrifice child
[  +0.006717] Killed process 19010 (rsyslogd) total-vm:3475172kB, anon-rss:1495608kB, file-rss:0kB, shmem-rss:0kB
[  +0.113619] oom_reaper: reaped process 19010 (rsyslogd), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
[Dec 4 04:28] Process accounting resumed

[  +0.008061] Out of memory: Kill process 16012 (rsyslogd) score 727 or sacrifice child
[  +0.007863] Killed process 16012 (rsyslogd) total-vm:3606236kB, anon-rss:1497316kB, file-rss:0kB, shmem-rss:0kB
[  +0.108939] oom_reaper: reaped process 16012 (rsyslogd), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
[Dec 5 04:44] Process accounting resumed

### Steps to reproduce the behavior
Happens sometimes.

### Environment
- rsyslog version: 
rsyslog-gnutls-8.1911.0-1.el6.x86_64
rsyslog-openssl-8.1911.0-1.el6.x86_64
rsyslog-8.1911.0-1.el6.x86_64

- platform: 
cat /etc/os-release
NAME="Amazon Linux AMI"
VERSION="2018.03"
ID="amzn"
ID_LIKE="rhel fedora"
VERSION_ID="2018.03"
PRETTY_NAME="Amazon Linux AMI 2018.03"
ANSI_COLOR="0;33"
CPE_NAME="cpe:/o:amazon:linux:2018.03:ga"
HOME_URL="http://aws.amazon.com/amazon-linux-ami/"

- for configuration questions/issues, include rsyslog.conf and included config files
 module(load="imuxsock") # provides support for local system logging (e.g. via logger command)
 module(load="imklog") # provides kernel logging support (previously done by rklogd)
module(load="imudp" SchedulingPolicy="fifo" SchedulingPriority="5" threads="2" timeRequery="8" batchSize="128")
input(type="imudp" port="514")
module(load="imptcp") # needs to be done just once
input(type="imptcp" port="514")

$MaxOpenFiles 102400

$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat

$template snippedFileFormat, "%$year% %timegenerated% %HOSTNAME% %msg%\n"
$template snippedXDPFileFormat, "%msg%\n"
$template PanFileFormat, "%HOSTNAME%,%msg%\r\n"
$template VormetricFF, "%$year% %timegenerated% %HOSTNAME% %syslogtag% %structured-data%\n"

$IncludeConfig /etc/rsyslog.d/.conf
$IncludeConfig /etc/rsyslog.d/snipped/.conf

global(
defaultNetstreamDriverCAFile="/opt/snipped/etc/apps/proofpoint/certs/cis-eng-ca.pem"
defaultNetstreamDriverCertFile="/opt/snipped/etc/apps/proofpoint/certs/logforwarder.snipped.pem"
defaultNetstreamDriverKeyFile="/opt/snipped/etc/apps/proofpoint/certs/logforwarder.snipped.pem"
)
module(
load="imtcp"
StreamDriver.Name="ossl"
StreamDriver.mode="1"
StreamDriver.AuthMode="anon"
)
input(type="imtcp" port="10514")

if $programname == 'puppet-agent' then -/var/log/all/puppet
if $programname == 'puppet-master' then -/var/log/all/puppet
if $programname == 'puppet-user' then -/var/log/all/puppet

*.info;mail.none;authpriv.none;cron.none /var/log/messages

authpriv.* /var/log/secure

mail.* /var/log/maillog

cron.* /var/log/cron

.emerg :omusrmsg:
uucp,news.crit /var/log/spooler
local7.* /var/log/boot.log

auth.debug -/var/log/all/auth.log
authpriv.debug -/var/log/all/authpriv.log
cron.debug -/var/log/all/cron.log
daemon.debug -/var/log/all/daemon.log
ftp.debug -/var/log/all/ftp.log
kern.debug -/var/log/all/kern.log
lpr.debug -/var/log/all/lpr.log
mail.debug -/var/log/all/mail.log
news.debug -/var/log/all/news.log
syslog.debug -/var/log/all/syslog.log
user.debug -/var/log/all/user.log
uucp.debug -/var/log/all/uucp.log
local0.debug -/var/log/all/local0.log
local1.debug -/var/log/all/local1.log
local2.debug -/var/log/all/local2.log
local3.debug -/var/log/all/local3.log
local4.debug -/var/log/all/local4.log
local5.debug -/var/log/all/local5.log
local6.debug -/var/log/all/local6.log
local7.debug -/var/log/all/local7.log

if $programname == 'snipped-drupal' then /var/log/drupal/drupal.log

template (name="Trend" type="string" string="/var/log/trend/trend_ds.log")
if ( $source == 'localhost' or $fromhost-ip == '127.0.0.1' ) and $syslogfacility-text == 'local5' then
{
action(type="omfile" dynaFile="Trend" template="RSYSLOG_TraditionalFileFormat")
stop
}

**Comment by ionosphere80** [comment]
Do you have anything running health checks on the TLS listener port?

**Comment by mike1li** [comment]
> 
> 
> Do you have anything running health checks on the TLS listener port?

Yes there are constant/high rate AWS NLB health checks on tcp_10514. I think there should be some logic to handle loadbalancer health check which may lack tls/ssl negotiation. Thanks.

**Comment by ionosphere80** [comment]
I am using the same architecture. When the NLB opens and closes a TCP connection for the health check, the `rsyslog` process considers it an invalid TLS connection and generates a bunch of log messages. I added a filter to remove those messages from the actual logs. However, each NLB health check causes the `rsyslog` process to consume a bit more memory until the kernel kills it. The syslog input module or OpenSSL library contains a bug that isn't freeing memory after an invalid or failed TLS connection. If I get some time, I might test the GnuTLS library to determine if it has the same issue. However, I can't use the GnuTLS library in production because it contains a different bug.
